\documentclass[12pt,glossary]{dalthesis}

\usepackage{amsmath}
\usepackage{amsmath}
\usepackage{bm} % for bold math symbols
\usepackage{booktabs} % better tables
%\usepackage[left=1in,right=1in,top=1in,bottom=1in]{geometry} % margins
% for subfigures (requires caption, breaks knitr w/o subfloat defined below)
%\usepackage{caption,subcaption}
\usepackage{float}
\usepackage{graphicx} % obviously for graphics
%\usepackage{latexsym} % MBE template for some fonts
\usepackage{lineno} % line numbers
\usepackage{mathtools} % an extension to amsmath to fix bugs
\usepackage{multirow} % column cells that span multiple rows
\usepackage{natbib} % nicer references
%\usepackage[natbib=true,style=apa]{biblatex}
%\addbibresource{/home/jrm/scm/references.git/refs.bib}
\usepackage{paralist} % inline lists
%\usepackage[section]{placeins} % keep figures and table inside section
\usepackage{rotating} % for landscape tables
\usepackage{setspace} % for line spacing
% need subfloat b/c knitr's fig.subcap was built with deprecated subfig package
\newcommand{\subfloat}[2][need a sub-caption]{\subcaptionbox{#1}{#2}}
\usepackage[flushleft]{threeparttable} % description under table
\usepackage{tikz}
\usepackage{tikz-qtree}
\usepackage[noindentafter,tiny]{titlesec}
\titleformat{\subsection}{\itshape\small\bfseries}{\thesubsection}{1em}{}
\titlespacing{\section}{0pt}{6pt}{6pt}
\titlespacing{\subsection}{0pt}{5pt}{5pt}
\usepackage{verbatim} % comments

\usetikzlibrary{arrows}

% specialcell is for line breaks in table cells
\newcommand{\specialcell}[2][c]{\begin{tabular}[#1]{@{}c@{}}#2\end{tabular}}

\renewcommand{\figurename}{Fig.}

\begin{document}

<<setup,include=F>>=
library(ggplot2)
library(grid)
library(gtable)
library(knitr)
opts_chunk$set(fig.path='figures/',fig.align='center',fig.show='hold',cache=T,autodep=T) #$
options(formatR.arrow=TRUE,width=90)
setwd("/home/jrm/scm/thesis.git/")
@

\title{Improving the Reliability of Models of Molecular Evolution with modified likelihood, bootstrapping, and blah}
\author{Joseph R. Mingrone}

\phd
\defencemonth{February}
\defenceyear{2018}

% \dedicate{Optionally, the thesis can be dedicated to someone, and the student can enter the dedication content here.}

\frontmatter

\begin{abstract}[s]
  % The study of life hinges on a unifying explanation for genetic diversity, evolution.
  Describing the mechanisms of evolution at the protein and amino acid levels is useful for connecting genetic and functional change.  Papers describing implementations of the most widely used models of molecular evolution have nearly $10,000$ citations from research in a variety of fields from vaccine design to mammalian physiology.  The aim of this thesis is to build upon these models to improve their reliability.

  Before the detection of positive selection at individual amino acid sites, models of molecular evolution commonly use a sweeping, protein-wide likelihood ratio test for positive selection.  Due to statistical irregularity, the distributions of the likelihood ratio statistic for these sweeping tests may not be strictly justified and thresholds determined from the distributions can give larger than expected type I error rates.  In Chapter 2, we present a modified likelihood approach that can restore statistical regularity to give tractable likelihood ratio statistic distributions for mixture models of codon evolution.

  To detect positive selection at individual amino acid sites, most methods use an empirical Bayes approach.  After parameters of a Markov process of codon evolution are estimated via maximum likelihood, they are passed to Bayes formula to compute the posterior probability that a site evolved under positive selection.  A difficulty with this approach is that parameter estimates with large errors can negatively impact Bayesian classification.  In Chapter 3, we present a technique we call smoothed bootstrap aggregation to accommodate the uncertainty in parameter estimates.

  In Chapter 4 we do something earth shattering that transforms the world.
\end{abstract}

\printglossary
\glossary{name={LR},description={Likelihood Ratio}}
\glossary{name={SBA},description={Smoothed Bootstrap Aggregation}}

%\begin{acknowledgements}
%Thanks to all the little people who make me look tall.
%\end{acknowledgements}

\mainmatter

\chapter{Introduction}
Although Darwin could only abstractly understand the concept of a gene or protein, the tenets of his theory of evolution apply at the molecular level.  Whether an adaptation arises in morphology, as a new behaviour, or at the molecular level, individuals expressing a beneficial trait will, on average, pass down their genes more frequently to the next generation.  However, while the same underlying processes drive evolution at the molecular level, the approaches for studying morphological and behavioural adaptations generally do not apply.  \cite{Lewontin} wrote, ``It has proved remarkably difficult to get compelling evidence for changes in enzymes brought about by selection, not to speak of adaptive changes.''.  Exponential advancements in both computing power and genetic sequencing technologies in the following decades have opened new opportunities.  As such, models of protein evolution used to detect amino acids subjected to positive selection have become more sophisticated.
% \begin{figure}[h!]
%   \centering
%   \includegraphics[scale=0.35]{images/me_growth.png}
%   \caption[Growth curves relevant to molecular evolution.]{The growth of genetic data and computing power over the past 40 years.  The graph shows the total number of nucleotides in the European Nucleotide Sequence Database\cite{EMBL} (circles) and the number of transistors in computer processors (squares\cite{INTEL2008} and diamonds\cite{WP2008}).  Note the log scale on the y-axis.}
%   \label{fig:me_growth}
% \end{figure}

A fundamental theoretical concept used when modelling protein evolution is the neutral theory first proposed by Kimura \cite{K1968}.  Since the predictions of the theory have been widely validated by analysis of genetic data, it is viewed as a highly useful framework for understanding molecular evolution \cite{L1997}.  The theory asserts that the majority of genetic diversity results from the random fixation of selectively neutral mutations and advantageous mutations are rare.  This means that simply locating regions of genetic diversity is not an effective method for detecting molecular adaptation.  However, neutral theory provides predictions about the evolution of protein coding genes that can be used as the basis of a hypothesis test \cite{YB2000,Nea2007}. One set of predictions makes use of the redundancy of the genetic code \cite{K1968}.  Because of this redundancy, a nucleotide substitution within a codon can either result in a change in the amino acid product (nonsynonymous substitution) or no change in the amino acid (synonymous substitution). FIGURE HERE.  Under a strictly neutral model (i.e., no selective consequences to any change in an amino acid) the nonsynonymous (\emph{dN}) and synonymous (\emph{dS}) substitution rates would be approximately the same. This leads to a computable index of selection: the ratio of the rates of nonsynonymous to synonymous substitutions, often expressed as the Greek letter \(\omega\) (\emph{dN/dS}=\(\omega\)).  As the two rates are equal under strict neutrality, the ratio is expected to be approximately $1$. However, neutral theory does not predict all amino acid changes are selectively neutral.  It predicts that the observable divergence between lineages was the result of a neutral process.  Genetic variation with negative fitness consequences also arises, but these deleterious mutations are removed by purifying selection \cite{K1968}.  Hence, the theory predicts that functional genes will have an \(\omega<1\) due to purifying selection pressure on deleterious, nonysynonymous changes \cite{K1968}.  The relevant null hypothesis then, is that the \(\omega\) ratio will be less than or equal to $1$ in a protein subject only to purifying selection pressure. The alternative hypothesis of molecular adaptation is characterized by an \(\omega\) ratio greater than $1$.  Nonsynonymous substitutions are fixed at a rate greater than synonymous substitutions (i.e., \(\omega>1\)) when they are advantageous and fixed due to positive selection.

%\subsection*{Counting and Fixed-Effect Methods}
Two general categories of methods for detecting positively selected amino acid sites include counting and fixed-effect methods.  Counting methods employ ancestral reconstruction of codon states for all internal nodes of a phylogenetic tree to obtain counts of the synonymous and nonsynonymous changes along each of its branches.  The counts inferred for a given site are used to test if $\omega \neq 1$.  Some counting methods use parsimony \citep{fitch1997long,bush1999positive,suzuki1999method}, and others likelihood \citep{suzuki2004new,nielsen2002mapping,nielsen2002detecting,suzuki2004false,pond2005not} to infer the ancestral codon states.  The reconstructions are often similar, but under the likelihood approach, uncertainty about the inference can be summarized via the posterior probabilities of the ancestral states.  Thus, the parsimony based methods must assume that these uncertainties are irrelevant to the statistical test.  While this makes the approach attractive for very large datasets where reliable reconstructions can be obtained relatively quickly \citep{lemey2012counting}, widespread use is hindered by a lack of power when the level of divergence is too low or by the negative impact of substitutional saturation when the level of divergence is too high \citep{pond2005not}.

An alternative approach is to treat each site as independently relevant to the question of evolution by positive selection, and attempt to fit an $\omega$ parameter to the data at each site.  Thus, the effect of each site on the task of $\omega$ inference is fixed.  Model based testing for $\omega \neq 1$ can be carried out via a standard likelihood ratio test (LR), and no assumptions are required about the distribution of selection pressure, $\omega$.  Although $\omega$ is treated as a site-specific variable, other important variables in the codon model (e.g., branch lengths) are shared among sites, with their values estimated jointly from the complete set of sites.  Results obtained by using these modelling ideas \citep{pond2005not, massingham2005detecting} are encouraging \citep{scheffler2014validity}, however, $\chi^2$ approximations to the distribution of the test statistic assume relatively large numbers of taxa, which is often not the case.  The lack of independence of data across taxa due to homology creates further difficulties for $\chi^2$ approximations.

% \subsection*{Empirical Bayes Methods}
A third approach for detecting positive selection at amino acid sites, which is the focus of this thesis, treats the value of \(\omega\) at a site as the realized value of a random variable.  From an alignment of protein coding DNA sequences, the \(\omega\) ratio can be estimated using maximum likelihood under an explicit model of codon evolution \cite{BY2005}.  However, because proteins subject to positive selection must still maintain the capacity to fold into complex structural and functional domains\cite{YNGP2000}, most amino acid sites will still be under purifying selection pressure. Only a small fraction will be subject to adaptive change and exhibit an \(\omega\) ratio \(>\) $1$. This means a single \(\omega\) ratio averaged over the entire sequence of such a protein would be less than $1$ and positive selection would go undetected. This problem is avoided by modeling several codon-site classes subject to different levels of selection pressure (e.g., \(\omega<1\), \(\omega=1\), \(\omega>1\)) \cite{NY1998}.  A Bayesian method can then be used to calculate a posterior probability (PP) that a given codon-site evolved under one of the site classes in the model.  Additionally, a posterior mean (PM) \(\omega\) can be computed as a measure of the selection pressure at each site.  Yang and colleagues introduced and implemented Markov models of codon evolution \cite{GY1994,NY1998} that have been widely used to accomplish these tasks.

MERGE PARAGRAPH BELOW INTO PARAGRAPH ABOVE

Identifying positively selected amino acid sites is a challenging statistical task that is useful for investigating the functional consequences of molecular change \citep{yang2005power}.  Several approaches have been developed to detect positive selection within a protein \citep[reviewed in][]{pond2005not,anisimova2009investigating}, but their reliability varies according to the properties of the data in hand.  The most widely used methods employ a codon model to detect an excess in the rate of nonsynonymous substitutions relative to synonymous substitutions ($dN/dS = \omega > 1$), which is an indication of evolution by positive selection.  Proteins evolving under positive selection must retain the capacity to fold into complex structural and functional domains, so the majority of amino acid substitutions will be subject to purifying selection pressure, with $\omega < 1$ \citep{kimura1968evolutionary}.  From extensive surveys of positive selection in real genes, we expect that only a small fraction of amino acid sites will be subject to adaptive change and exhibit an $\omega > 1$ \citep[e.g., ][]{anisimova2007phylogenomic,ge2008protein}.  The sparseness of these sites makes them challenging to identify.


There are recommendations \citep[e.g.,][]{yang1998synonymous} to use a pre-screen that fits two models: one with a distribution that excludes values of \(\omega>1\), and another with the same distribution, except with weight on values of \(\omega>1\) permitted.  This nested-model pre-screening is used to test if the data conveys any evidence of positive selection.  When the null hypothesis of no positive selection is rejected using a LR test, site-wise analysis is warranted.  Site-wise analysis is carried out using Bayes rule to calculate the posterior probability that a site \(h\) evolved under some estimated value of \(\omega\), given the data at site \(h\).  This approach is referred to as empirical Bayes (EB) because the marginal distribution of \(\omega\) is determined from the data.  Conclusions regarding the evolution at a site are made based on the estimated \(\omega\)-values along with their associated posterior probabilities conditioned on the data at the site.  For example, when the largest posterior probability for a site is associated with a value of \(\omega>1\), this is taken as evidence of positive selection at that site.

\chapter{Modified Likelihood}
\section{Introduction}

\chapter{Smoothed Bootstrap Aggregation}
\section{Introduction}

\chapter{Third Project}
\section{Introduction}

\bibliographystyle{plain}
\bibliography{/home/jrm/scm/references.git/refs}

\end{document}
